testonly/integration/clientserver.go

```diff
@@ -115,62 +115,95 @@ func getTestDB(testID string) (*sql.DB, error) {
 	return dbTest, nil
 }
 
-// NewLogEnv creates a fresh DB, log server, and client.
+// NewLogEnv creates a fresh DB, log server, and client. The numSequencers parameter
+// indicates how many sequencers to run in parallel; if numSequencers is zero a
+// manually-controlled test sequencer is used.
 // testID should be unique to each unittest package so as to allow parallel tests.
-func NewLogEnv(testID string) (*LogEnv, error) {
+func NewLogEnv(ctx context.Context, numSequencers int, testID string) (*LogEnv, error) {
 	db, err := getTestDB(testID)
 	if err != nil {
 		return nil, err
 	}
 
 	timesource := &util.SystemTimeSource{}
-	registry, err := builtin.NewDefaultExtensionRegistry()
+	registry, err := builtin.NewExtensionRegistry(db)
 	if err != nil {
 		return nil, err
 	}
 
-	// Start Log Server.
+	// Create Log Server.
 	grpcServer := grpc.NewServer()
 	logServer := server.NewTrillianLogRPCServer(registry, timesource)
 	trillian.RegisterTrillianLogServer(grpcServer, logServer)
 
-	// Start Sequencer.
+	// Create Sequencer.
 	keyManager, err := crypto.LoadPasswordProtectedPrivateKey(privateKeyFile, privateKeyPassword)
 	if err != nil {
 		return nil, err
 	}
-	sequencerManager := server.NewSequencerManager(keyManager, registry,
-		sequencerWindow)
-	sequencerTask := server.NewLogOperationManagerForTest(ctx, registry,
-		batchSize, sleep, timeSource, sequencerManager)
+	sequencerManager := server.NewSequencerManager(keyManager, registry, sequencerWindow)
+	var wg sync.WaitGroup
+	var sequencerTask *server.LogOperationManager
+	var cancel context.CancelFunc
+	if numSequencers == 0 {
+		// Test sequencer that needs manual triggering (with env.Sequencer.OperationLoop()).
+		sequencerTask = server.NewLogOperationManagerForTest(ctx, registry,
+			batchSize, sleepBetweenRuns, timeSource, sequencerManager)
+	} else {
+		// Start a live sequencer in a goroutine.
+		var ctx2 context.Context
+		ctx2, cancel = context.WithCancel(ctx)
+		sequencerTask = server.NewLogOperationManager(ctx2, registry,
+			batchSize, numSequencers, sleepBetweenRuns, timesource, sequencerManager)
+		wg.Add(1)
+		go func(wg *sync.WaitGroup, om *server.LogOperationManager) {
+			defer wg.Done()
+			om.OperationLoop()
+		}(&wg, sequencerTask)
+	}
 
 	// Listen and start server.
 	addr, lis, err := listen()
 	if err != nil {
+		if cancel != nil {
+			cancel()
+		}
 		return nil, err
 	}
-	go grpcServer.Serve(lis)
+	wg.Add(1)
```

Comment:
`log.Close()` already calls `grpcServer.GracefulStop()`, so I'm not sure what you're after here.

I think a wait group is absolutely essential when we starting separate goroutines -- for example, to ensure that any shutdown failure shows up as a hang/test timeout rather than an obscure database locking issue.